Finalize implementation of MediaStreamTrack powerEfficient constraint

youennf opened the pull request:
#### 312f06335038c0c43733c76223f9a53806a002c1
<pre>
Finalize implementation of MediaStreamTrack powerEfficient constraint
<a href="https://bugs.webkit.org/show_bug.cgi?id=276075">https://bugs.webkit.org/show_bug.cgi?id=276075</a>
<a href="https://rdar.apple.com/130897389">rdar://130897389</a>

Reviewed by Eric Carlson.

We were previoudly not exposing powerEfficient settings and capabilities.
Now that the spec changes have landed in <a href="https://w3c.github.io/mediacapture-extensions/#the-powerefficient-constraint">https://w3c.github.io/mediacapture-extensions/#the-powerefficient-constraint</a>,
we can expose them.
We only set powerEfficient to [false, true] if there are presets that are power efficients and others that are not.
Otherwise we set it to [false].

* LayoutTests/fast/mediastream/camera-powerEfficient-track-expected.txt:
* LayoutTests/fast/mediastream/camera-powerEfficient-track.html:
* Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
(WebCore::MediaStreamTrack::getSettings const):
* Source/WebCore/Modules/mediastream/MediaStreamTrack.h:
* Source/WebCore/Modules/mediastream/MediaStreamTrack.idl:
* Source/WebCore/Modules/mediastream/MediaTrackCapabilities.cpp:
(WebCore::powerEfficientCapabilityVector):
(WebCore::toMediaTrackCapabilities):
* Source/WebCore/Modules/mediastream/MediaTrackCapabilities.h:
* Source/WebCore/Modules/mediastream/MediaTrackCapabilities.idl:
* Source/WebCore/platform/mediastream/RealtimeMediaSourceCapabilities.h:
(WebCore::RealtimeMediaSourceCapabilities::RealtimeMediaSourceCapabilities):
(WebCore::RealtimeMediaSourceCapabilities::supportsPowerEfficient const):
(WebCore::RealtimeMediaSourceCapabilities::powerEfficient const):
(WebCore::RealtimeMediaSourceCapabilities::setPowerEfficient):
(WebCore::RealtimeMediaSourceCapabilities::isolatedCopy const):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceSettings.cpp:
(WebCore::RealtimeMediaSourceSettings::isolatedCopy const):
(WebCore::RealtimeMediaSourceSettings::convertFlagsToString):
(WebCore::RealtimeMediaSourceSettings::difference const):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceSettings.h:
(WebCore::RealtimeMediaSourceSettings::allFlags):
(WebCore::RealtimeMediaSourceSettings::RealtimeMediaSourceSettings):
(WebCore::RealtimeMediaSourceSettings::supportsPowerEfficient const):
(WebCore::RealtimeMediaSourceSettings::powerEfficient const):
(WebCore::RealtimeMediaSourceSettings::setPowerEfficient):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceSupportedConstraints.h:
(WebCore::RealtimeMediaSourceSupportedConstraints::RealtimeMediaSourceSupportedConstraints):
(WebCore::RealtimeMediaSourceSupportedConstraints::supportsPowerEfficient const):
(WebCore::RealtimeMediaSourceSupportedConstraints::setSupportsPowerEfficient):
* Source/WebCore/platform/mediastream/RealtimeVideoCaptureSource.cpp:
(WebCore::RealtimeVideoCaptureSource::canBePowerEfficient):
* Source/WebCore/platform/mediastream/RealtimeVideoCaptureSource.h:
* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
(WebCore::AVVideoCaptureSource::settings):
(WebCore::AVVideoCaptureSource::capabilities):
* Source/WebCore/platform/mock/MockRealtimeVideoSource.cpp:
(WebCore::MockRealtimeVideoSource::capabilities):
(WebCore::MockRealtimeVideoSource::settings):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::isMediaStreamTrackPowerEfficient const): Deleted.
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
(WebKit::UserMediaCaptureManagerProxy::createMediaSourceForCaptureDeviceWithConstraints):
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSource.cpp:
(WebKit::RemoteRealtimeMediaSource::createRemoteMediaSource):

Canonical link: <a href="https://commits.webkit.org/280576@main">https://commits.webkit.org/280576@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b3dded311bd97a6a8dbba2c6212c3e45fde6377f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/56998 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36326 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9473 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60618 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7441 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/59126 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/43949 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7631 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46163 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5228 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59028 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34112 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49195 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27023 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/30892 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6526 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6446 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/52853 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6797 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62299 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/911 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/6902 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53420 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/914 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49250 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53459 "Found 1 new API test failure: /WebKitGTK/TestWebViewEditor:/webkit/WebKitWebView/select-all/non-editable (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12605 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/772 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32155 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33240 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34325 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/32986 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/Modules/mediastream/MediaTrackCapabilities.cpp:
```diff
@@ -132,6 +143,11 @@ MediaTrackCapabilities toMediaTrackCapabilities(const RealtimeMediaSourceCapabil
     if (capabilities.supportsBackgroundBlur())
         result.backgroundBlur = capabilityBooleanVector(capabilities.backgroundBlur());
 
+    fprintf(stderr, "toMediaTrackCapabilities %d %d\n", capabilities.supportsPowerEfficient(), capabilities.powerEfficient());
```
Nit: don’t forget to remove this

youennf replied:
Done

eric-carlson commented on Source/WebCore/platform/mock/MockRealtimeVideoSource.cpp:
```diff
@@ -244,6 +244,12 @@ const RealtimeMediaSourceCapabilities& MockRealtimeVideoSource::capabilities()
+        capabilities.setPowerEfficient(canBePowerEfficient());
+        supportedConstraints.setSupportsPowerEfficient(true);
+
+        fprintf(stderr, "MockRealtimeVideoSource::capabilities %d\n", capabilities.powerEfficient());
```
Ditto

eric-carlson commented on Source/WebCore/platform/mock/MockRealtimeVideoSource.cpp:
```diff
@@ -365,6 +371,12 @@ const RealtimeMediaSourceSettings& MockRealtimeVideoSource::settings()
             settings.setTorch(torch());
         }
 
+        fprintf(stderr, "MockRealtimeVideoSource::settings %d %d\n", !!m_preset, m_preset ? m_preset->isEfficient() : false);
```
Ditto

youennf replied:
Done

eric-carlson commented on Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
```diff
@@ -558,6 +558,9 @@ void UserMediaCaptureManagerProxy::createMediaSourceForCaptureDeviceWithConstrai
             return;
         }
 
+        fprintf(stderr, "UserMediaCaptureManagerProxy::createMediaSourceForCaptureDeviceWithConstraints %d %d\n", proxy->source().capabilities().supportsPowerEfficient(), proxy->source().capabilities().powerEfficient());
```
Ditto

eric-carlson commented on Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSource.cpp:
```diff
@@ -58,6 +58,8 @@ void RemoteRealtimeMediaSource::createRemoteMediaSource()
             return;
         }
 
+        fprintf(stderr, "RemoteRealtimeMediaSource::createRemoteMediaSource %d %d\n", capabilities.supportsPowerEfficient(), capabilities.powerEfficient());
```
Ditto

youennf replied:
Done

youennf committed:
Finalize implementation of MediaStreamTrack powerEfficient constraint
https://bugs.webkit.org/show_bug.cgi?id=276075
rdar://130897389

Reviewed by Eric Carlson.

We were previoudly not exposing powerEfficient settings and capabilities.
Now that the spec changes have landed in https://w3c.github.io/mediacapture-extensions/#the-powerefficient-constraint,
we can expose them.
We only set powerEfficient to [false, true] if there are presets that are power efficients and others that are not.
Otherwise we set it to [false].

* LayoutTests/fast/mediastream/camera-powerEfficient-track-expected.txt:
* LayoutTests/fast/mediastream/camera-powerEfficient-track.html:
* Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
(WebCore::MediaStreamTrack::getSettings const):
* Source/WebCore/Modules/mediastream/MediaStreamTrack.h:
* Source/WebCore/Modules/mediastream/MediaStreamTrack.idl:
* Source/WebCore/Modules/mediastream/MediaTrackCapabilities.cpp:
(WebCore::powerEfficientCapabilityVector):
(WebCore::toMediaTrackCapabilities):
* Source/WebCore/Modules/mediastream/MediaTrackCapabilities.h:
* Source/WebCore/Modules/mediastream/MediaTrackCapabilities.idl:
* Source/WebCore/platform/mediastream/RealtimeMediaSourceCapabilities.h:
(WebCore::RealtimeMediaSourceCapabilities::RealtimeMediaSourceCapabilities):
(WebCore::RealtimeMediaSourceCapabilities::supportsPowerEfficient const):
(WebCore::RealtimeMediaSourceCapabilities::powerEfficient const):
(WebCore::RealtimeMediaSourceCapabilities::setPowerEfficient):
(WebCore::RealtimeMediaSourceCapabilities::isolatedCopy const):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceSettings.cpp:
(WebCore::RealtimeMediaSourceSettings::isolatedCopy const):
(WebCore::RealtimeMediaSourceSettings::convertFlagsToString):
(WebCore::RealtimeMediaSourceSettings::difference const):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceSettings.h:
(WebCore::RealtimeMediaSourceSettings::allFlags):
(WebCore::RealtimeMediaSourceSettings::RealtimeMediaSourceSettings):
(WebCore::RealtimeMediaSourceSettings::supportsPowerEfficient const):
(WebCore::RealtimeMediaSourceSettings::powerEfficient const):
(WebCore::RealtimeMediaSourceSettings::setPowerEfficient):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceSupportedConstraints.h:
(WebCore::RealtimeMediaSourceSupportedConstraints::RealtimeMediaSourceSupportedConstraints):
(WebCore::RealtimeMediaSourceSupportedConstraints::supportsPowerEfficient const):
(WebCore::RealtimeMediaSourceSupportedConstraints::setSupportsPowerEfficient):
* Source/WebCore/platform/mediastream/RealtimeVideoCaptureSource.cpp:
(WebCore::RealtimeVideoCaptureSource::canBePowerEfficient):
* Source/WebCore/platform/mediastream/RealtimeVideoCaptureSource.h:
* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
(WebCore::AVVideoCaptureSource::settings):
(WebCore::AVVideoCaptureSource::capabilities):
* Source/WebCore/platform/mock/MockRealtimeVideoSource.cpp:
(WebCore::MockRealtimeVideoSource::capabilities):
(WebCore::MockRealtimeVideoSource::settings):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::isMediaStreamTrackPowerEfficient const): Deleted.
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
(WebKit::UserMediaCaptureManagerProxy::createMediaSourceForCaptureDeviceWithConstraints):
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSource.cpp:
(WebKit::RemoteRealtimeMediaSource::createRemoteMediaSource):

Canonical link: https://commits.webkit.org/280576@main
